doc/design.md

```diff
@@ -1,51 +1,203 @@
 
-It is not a problem that both folders overlap. In case a file is defined in both directories, 
-this plugin validates it as if it would only exist in `require_exact`. 
+We decided for this interface since it allows us to:
+
+* Report errors that can't be fixed automatically
+* Report error and directly provide a fix for it
+
+A problem with this approach is that some validation-fixes change a resource (typically files). If a validator creates multiple validation findings for the same resource, each fix needs to read and write the file. Consider for example the plugin validator for the pom file. If the pom file validator detects multiple missing plugins, each of the fixes would need to write the file. Since the fixes are executed by the surrounding code and not in a specific order. To solve this, we used the trick to create a compound finding. So basically one finding with one error message that combines the one of all it's sub findings. Its fix is a method that runs the fixes of all sub-findings and then writes the pom file.
 
 Covers:
 
-* `req~repository-structure~1`
-* `req~convenient-repository-structure~1`
+* feat~fix-findings~1
 
 Needs: impl
 
-## Maven Verify Goal
+### Required Files Validator
+
+`dsn~required-files-validator~1`
+
+We decided to add the template for the required files by a file structure in the resources of this project. There are two folders: `require_exist` and `require_exact`. For files in the `require_exist` folder, PK only checks if they exist in the project. If they don't exist it creates the file with the content from the file in `require_exist` as a template.
+
+For the files in the `require_exact` PK also validates that the content is exactly the same.
+
+Rationale:
+This way of defining the project structure makes it very convenient to add new content. No knowledge about the internal implementation of this plugin is required. That allows users to quickly add or modify the template.
+
+An alternative would be to define the files by Java classes. In order to contain the contents of the required files, these classes would contain many String concatenations which would be very unreadable.
+
+Covers:
+
+* `req~verify-existence-of-files~1`
+
+Needs: impl, utest, itest
+
+### Deleted Files Validator
+
+`dsn~deleted-files-validator~1`
+
+For configuring the deleted files validator we decided for a Java API. For each deleted file one can specify a reason that will be printed as error message. By that one can for example tell that the file was moved.
+
+Covers:
+
+* `req~verify-non-existence-of-files~1`
+
+Needs: impl, utest, itest
+
+### Maven Plugin Validator
+
+`dsn~mvn-plugin-validator~1`
+
+We decided to implement the verification of the pom plugins using a combination of xml files and Java code. The XML files (in resources/mvn_templates) contain the XML of a typical plugin definition. The Java class defines which XML properties must be like define in the template.
```

Comment:
```suggestion
We decided to implement the verification of the pom plugins using a combination of XML files and Java code. The XML files (located in `resources/mvn_templates` directory) contain the XML of a typical plugin definition. The Java class defines which XML properties must be like define in the template.
```

Reply:
This is not  very clear: The Java class defines which XML properties must be like define in the template.

Did you mean, `The Java class verifies that XML properties defined in the pom file are same as the ones in the template.` ?

Reply:
```suggestion
We decided to implement the verification of the pom plugins using a combination of XML files and Java code. The XML files (located in `resources/mvn_templates` directory) contain the XML of a typical plugin definition. The Java class defines for which XML properties PK verifies that they are same as the ones in the template.
```

Reply:
Looks good!